interpreter/decorators.go

```diff
@@ -118,6 +121,61 @@ func decOptimize() InterpretableDecorator {
 	}
 }
 
+// decCompileRegexConstants looks for regex pattern string constants, compiles any it finds and returns any
+// regex pattern compilation errors encountered.
+func decCompileRegexConstants() InterpretableDecorator {
```

Comment:
Consider instead a decorator which produces an alternative `InterpretableCall`:

```go
struct regexMatcher {
  function string
  regexIndex int
  factory func(call interpreter.InterpretableCall, regexIndex int, pattern ref.Val) (interepreter.Interpretable, error)
}

func regexOptimizer(matchers ...regexMatcher) interpreter.InterpretableDecorator {
       regexMatchMap := make(map[string]int)
       for _, m := range matchers {
          regexMatchMap[m.function] = regexMatcher
       }
       return func(i interpreter.Interpretable) (interpreter.Interpretable, error) {
            call, ok := i.(interpreter.InterpretableCall)
            if !ok {
              return i, nil
            }
            matcher, found := regexMatchMap[call.Function()]
            if !found || matcher.regexIndex >= len(call.Args()) {
              return i, nil
            }
            args := call.Args()
            regexArg := args[matcher.regexIndex]
            regexStr, isConst := regexArg.(interpreter.InterpretableConst)
            if !isConst {
              return i, nil
            }
            return matcher.factory(call, matcher.regexIndex, regexStr.Value())
        }
    }
}
```

I would then expect the specific optimizers to look something like the following:

```go
regexMatcher{
    function: "matches", 
    regexIndex: 1, 
    factory: func(call interpreter.InterpretableCall, regexIndex int, pattern ref.Val) (intepreter.Interpretable, error) {
        compiled, err := regexp.Compile(pattern)
	if err != nil {
	    return nil, fmt.Errorf("%v in '%s' function call", err, call.Function())
	}
        return interpreter.NewCall(call.ID(), call.Function(), call.OverloadID(), args[1:2], 
            func(params ...ref.Val) ref.Val {
                in, ok := params[0].(String)
                if !ok {
                  return types.MaybeNoSuchOverloadErr(params[0])
                }
                return types.Bool(compiled.Match(string(in)))
            }), nil
    },
}
```

I'm slightly oversimplifying, but not by much. The goal would be to minimize the top-level API changes. That said, you'd need to expose an `interpreter.NewInterpretableCall` (or similar) to produce a program plan step which would replace the original call and implementation.

It may not look very nice, but it lets you provide your own optimizations which should also be something you can enable without much impact on the exposed APIs

Reply:
SGTM. I've restructured to match. This no longer provides any default optimizations, but there is an example of how to use it in the tests for "matches".